New pl-matching element and docs (#3670)
PrairieLearn/PrairieLearn: docs/elements.md

  <pl-answer match="d.c.">United States</pl-answer>
  <pl-answer match="Mexico City">Mexico</pl-answer>
  <pl-answer match="Paris">France</pl-answer>

  <pl-option name="d.c.">Washington, D.C.</pl-option>
  <pl-option>Mexico City</pl-option>
  <pl-option>Paris</pl-option>
</pl-matching>
```

#### Customizations

Attribute | Type | Default | Description
--- | --- | --- | ---
`answers-name` | string | — | Variable name to store data in.
`weight` | integer | 1 | Weight to use when computing a weighted average score over elements.
`fixed-order` | boolean | False | Whether or not to display the answers in a fixed order; otherwise they are shuffled. Options are always shuffled.
`number-answers` | integer | special | The number of answers to display. Defaults to all answers.
`number-options` | integer | special | The number of options to display. Defaults to all options.
`none-of-the-above` | boolean  | false | Whether or not to add a "None of the above" to the end of the options. Automatically set to true if number-options is less than the total number of options.
`blank` | boolean | False | Option to add blank dropdown entry as the default selection in each drop-down list..

> `counter-type` | "decimal" or "lower-alpha" or "upper-alpha" | "decimal" | The type of counter to use when enumerating the options.

Reviewer: I'd love to see an option where what's shown is the option text itself (either the name or the text). This can be useful when there is a limited number of options. For example, I have a question where students match to "static" or "dynamic", and listing these instead of 1/2 or a/b is probably less error-prone to students. I don't know if this would be a separate option (e.g., `include-option-text` or something like that) or replace the counter in that case, let me know what you think.

-----

### `pl-matching` element

Given a list of statements, select a matching option for each entry from a drop-down list.

#### Sample element

![](elements/pl-matching.png)

**question.html**
```html
<pl-matching answers-name="string_value">
  <pl-statement match="d.c.">United States</pl-statement>
  <pl-statement match="Mexico City">Mexico</pl-statement>
  <pl-statement match="Paris">France</pl-statement>

  <pl-option name="d.c.">Washington, D.C.</pl-option>

>   <pl-option>Mexico City</pl-option>
>   <pl-option>Paris</pl-option>

Reviewer: ```suggestion
  <pl-option>New York City</pl-option>
```

#### Customizations

Attribute | Type | Default | Description
--- | --- | --- | ---
`answers-name` | string | — | Variable name to store data in.
`weight` | integer | 1 | Weight to use when computing a weighted average score over elements.
`fixed-order` | boolean | False | Whether or not to display the statements in a fixed order; otherwise they are shuffled. Options are always shuffled.
`number-statements` | integer | special | The number of statements to display. Defaults to all statements.

> `number-options` | integer | special | The number of options to display. Defaults to all options.

Reviewer: ```suggestion
`number-options` | integer | special | The number of options to display. Defaults to all options. The `none-of-the-above` option does not count towards this number.
```

That is true, right?

`number-statements` | integer | special | The number of statements to display. Defaults to all statements.
`number-options` | integer | special | The number of options to display. Defaults to all options.
`none-of-the-above` | boolean  | false | Whether or not to add a "None of the above" to the end of the options.

> `blank` | boolean | False | Option to add blank dropdown entry as the default selection in each drop-down list..

Reviewer: ```suggestion
`blank` | boolean | False | Option to add blank dropdown entry as the default selection in each drop-down list.
```

I'm wondering also if it should default to True instead, which is the default behaviour of `pl-dropdown`. What do you think?

Author: I'm fine with that.

Inside the `pl-matching` element, a series of `pl-statement` and `pl-option` elements specify the questions the student must answer and the options to which they can be matched, respectively. Statements are displayed in the left column, and options in the right.

> A total of `number-statements` statements will be randomly selected and displayed to the student. The corresponding matching options will be gathered; if `number_options` requires more entries, then distractors will be selected from the remaining unused options. If the gathered options are more numerous than `number_options`, they will be randomly sampled and `none-of-the-above` will automatically set to true.

Reviewer: ```suggestion
A total of `number-statements` statements will be randomly selected and displayed to the student. The corresponding matching options will be gathered; if `number-options` is larger than the number of options used by the selected statements, then random distractors will be selected from the remaining unused options. If the selected statements require more options than `number-options`, then `none-of-the-above` will automatically set to true.
```

A total of `number-statements` statements will be randomly selected and displayed to the student. The corresponding matching options will be gathered; if `number_options` requires more entries, then distractors will be selected from the remaining unused options. If the gathered options are more numerous than `number_options`, they will be randomly sampled and `none-of-the-above` will automatically set to true.

The content of a `pl-statement` can be any HTML element, including other PrairieLearn elements. A `pl-statement` must be specified with these attributes:

Attribute | Type | Default | Description
--- | --- | --- | ---

> `match` | string | — | Must be equal to the `name` of one of the `pl-option` elements, and identifies that option as the correct response for this `pl-answer.`

Reviewer: ```suggestion
`match` | string | — | Identifies that option as the correct response for this `pl-answer`. If `match` corresponds to the `name` of any `pl-option` element, the statement will be linked to that `pl-option`, otherwise a new option is implicitly created based on this `match` value.
```

> The content of a `pl-option` can be any HTML element, including other PrairieLearn eleemnts. `A pl-option` must be specified with these attributes:

Reviewer: ```suggestion
The content of a `pl-option` can be any HTML element, including other PrairieLearn elements. Including `pl-option` tags is optional, since options are created by default based on the `match` setting of each `pl-statement`;   additional `pl-option` tags can be added to serve as distractors (an option that is always incorrect, such as "New York City" in the example above), or to provide a more detailed description of the option (such as "d.c." in the example above). A `pl-option` may specified with these attributes:
```

> `A pl-option` must be specified with these attributes:

Reviewer: ```suggestion
A `pl-option` must be specified with these attributes:
```
